share/p2p/peers/manager.go

```diff
@@ -0,0 +1,225 @@
+package peers
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/celestiaorg/celestia-node/header"
+	libhead "github.com/celestiaorg/celestia-node/libs/header"
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/availability/discovery"
+	"github.com/celestiaorg/celestia-node/share/p2p/shrexsub"
+)
+
+// TODO: Needs improvement:
+//   - make params configurable
+//   - add metrics, traces
+
+var (
+	log         = logging.Logger("shrex/peers")
+	syncTimeout = 60 * time.Second
+)
+
+// Manager keeps track of peers coming from shrex.Sub and discovery
+type Manager struct {
+	disc      discovery.Discovery
+	headerSub libhead.Subscription[*header.ExtendedHeader]
+
+	m         *sync.Mutex
+	pools     map[hashStr]syncPool
+	fullNodes pool
+
+	cancel context.CancelFunc
+	done   chan struct{}
+}
+
+type hashStr string
```

Comment:
I think we should just use string, and call it dataHash/hash when used as a parameter

Reply:
I like the second suggestion more. If caller always has access to share.DataHash directly, that the way to go.